proposal: TSDB Support for Start Timestamp (ST) (#60)
prometheus/proposals: proposals/0060-ct-storage.md

## Why

The main goal of this proposal is to make sure [PROM-29's created timestamp (CT)](0029-created-timestamp.md) information is reliably and efficiently stored in Prometheus TSDB, under the new chosen start timestamp (ST) name. This means ST can be:

* Written via TSDB Appender interfaces.
* Query-able via TSDB Querier interfaces.
* Persistent in WAL.
* Watch-able (WAL) by Remote Writer.
* (eventually) Persistent in TSDB block storage.

To do it reliably, we propose to extend TSDB storage to "natively" support ST as something you can attach to a sample and use later on.
Native ST support in Prometheus TSDB would unblock the practical use of ST information for:

* Remote storages (Remote Write 2.0) (e.g. OpenTelemetry, Chronosphere, Google).
* PromQL and other read APIs (including federation) (e.g. increased cumulative based operation accuracy).

Furthermore, it would unblock future Prometheus features for wider range of monitoring cases like:

* Delta temporality support.
* UpAndDown counter (i.e. not monotonic counters) e.g. StatsD.

> On top of that this work also allow to improve some existing features, notably various edge cases for native histogram resets hints (e.g. [TSDB chunk merges](https://github.com/prometheus/prometheus/issues/15346)). Arguably, ST can replace resent hint field in native histogram eventually.

Author: done, thanks!

4. Let's go through all the areas that would need to change:

### Feature flag `st-storage`

To develop and give experimental state to users, we propose to add a new feature flag `st-storage`. Similar to [exemplar-storage]https://prometheus.io/docs/prometheus/latest/feature_flags/#exemplars-storage) it will tell Prometheus to ingest ST and use new (potentially breaking compatibility) storage formats.

We propose to have a single flag for both WAL, Block storage, etc., to avoid tricky configuration.

Notably, given persistence of this feature, similar to example storage, if users enabled and then disabled this feature, users will might be able to access their STs through all already persistent pieces e.g. WAL).

This feature could be considered to be switched to opt-out, only after it's finished (this proposal is fully implemented) stable, provably adopted and when the previous LTS Prometheus version is compatible with this feature.

### Proposed ST semantics and validation

See the [official ST semantics in the ecosystem](#background-official-st-semantics). In this section we propose what semantics we recommend (or enforce).

> ST notion was popularized by OpenMetrics and OpenTelemetry. Early experience exposed a big challenge: ST data is extremely unclean given early adoption, mixed instrumentation support and multiple (all imperfect) ways of ["auto-generation"](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstarttimeprocessor) (`subtract_initial_point` might the most universally "correct", but it added only recently). This means that handling (or reducing ) ST errors is an important detail for consumers of this data.

Author: done, thanks!

│ ┌──────────────────────────────┬───────────────────────────────┬────────────────┐   │
│ │ ref_delta (vs prev) <varint> │ T_delta (vs first T) <varint> │ st_marker <1b> │   │
│ ├──────────────────────────────┴───────────────────────────────┴────────────────┤   │
│ │ [ST_delta (vs first ST) <varint>]  (only if st_marker = explicitST)           │   │
│ ├───────────────────────────────────────────────────────────────────────────────┤   │
│ │ value <8b>                                                                    │   │
│ └───────────────────────────────────────────────────────────────────────────────┘   │
│                                      . . .                                          │
└─────────────────────────────────────────────────────────────────────────────────────┘

ST marker values:
  0 (noST)       -- sample has no start timestamp (ST=0)
  1 (sameST)     -- same ST as previous sample
  2 (explicitST) -- explicit ST, followed by ST_delta (vs first ST)
```

Similar V2 records will be created for Histogram types.

#### TSDB blocks

ST will be persisted in TSDB blocks using a new chunk encoding currently under development.

#### Memory Snapshots and Head Chunks

> There will be likely some code to add to ensure STs are used properly, but the new proposed chunk format from [TSDB Block](#tsdb-blocks) section, should immidately work with the [Memory Snapshot](https://github.com/prometheus/prometheus/blob/747c5ee2b19a9e6a51acfafae9fa2c77e224803d/tsdb/docs/format/memory_snapshot.md) and [Head Chnks](https://github.com/prometheus/prometheus/blob/747c5ee2b19a9e6a51acfafae9fa2c77e224803d/tsdb/docs/format/head_chunks.md) formats.

Author: done, thanks!

  * start_timestamp SHOULD be provided for metrics that follow counter semantics (e.g. counters and histograms).
  * Receivers MAY reject those series without start_timestamp being set.

* Otel) More detailed, but still descriptive SHOULD rule only: In OpenTelemetry data model ([temporality section](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#sums:~:text=name%2Dvalue%20pairs.-,A%20time%20window%20(of%20(start%2C%20end%5D)%20time%20for%20which%20the,The%20time%20interval%20is%20inclusive%20of%20the%20end%20time.,-Times%20are%20specified)) ST is generally optional, but strongly recommended. Rules are also soft and we assume they are "SHOULD" because there is section of [handling overlaps](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#overlap). However, it provides more examples, which allow us to capture some specifics:
  * Time intervals are half open `(ST, T]`.
  * Generally, ST SHOULD:

>     * `ST[i] < T[i]` (equal means unknown)

Reviewer: does Prometheus have a general philosophy of "fix on ingest" versus "fix at query time"? From my own knowledge, I feel like usually Prometheus does "reject bad data on ingest" as its usual pattern. My concern is that if there's a mix of approaches, it will make Prometheus' behavior harder to reason about. 

As I think about it, I can't think of a problem with force-setting ST=0 if ST=T.

  * [initial attempt for ct per sample](https://github.com/prometheus/prometheus/pull/16046)
  * [rw2 proto change for ct per sample](https://github.com/prometheus/prometheus/pull/17036)
  * Initial implementation
    * [Appender](https://github.com/prometheus/prometheus/pull/17104)
    * [chunkenc.Iterator](https://github.com/prometheus/prometheus/pull/17176)

* **Other docs or links:**
  * [PROM-29 (Created Timestamp)](https://github.com/prometheus/proposals/blob/main/proposals/0029-created-timestamp.md)
  * [Delta type proposal](https://github.com/prometheus/proposals/pull/48), [Delta WG](https://docs.google.com/document/d/1G0d_cLHkgrnWhXYG9oXEmjy2qp6GLSX2kxYiurLYUSQ/edit)

> TL;DR: We propose to extend Prometheus TSDB storage sample definition to include an extra int64 that will represent the cumulative created timestamp (CT) and, for the future delta temporality ([PROM-48](https://github.com/prometheus/proposals/pull/48)), a delta start timestamp (ST).
> We propose introducing persisting CT logic behind a single flag `ct-storage`.

> > Once implemented, we propose to deprecate the `created-timestamps-zero-injection` experimental feature.

Reviewer 2: ```suggestion
> Once implemented, we propose to remove the `created-timestamps-zero-injection` experimental feature.
```

Do we need to go through a deprecation notice for feature flags? I have the impression that the point of feature flags is that they are safe to remove 🤔

Reviewer: experimental features esp have no guarantee of stability, so that seems correct

## Why

The main goal of this proposal is to make sure [PROM-29's created timestamp (CT)](0029-created-timestamp.md) information is reliably and efficiently stored in Prometheus TSDB, so:

* Written via TSDB Appender interfaces.
* Query-able via TSDB Querier interfaces.
* Persistent in WAL.
* Watch-able (WAL) by Remote Writer.
* (eventually) Persistent in TSDB block storage.

To do it reliably, we propose to extend TSDB storage to "natively" support CT as something you can attach to a sample and use later on.
Native CT support in Prometheus TSDB would unblock the practical use of CT information for:

* Remote storages (Remote Write 2.0) (e.g. OpenTelemetry, Chronosphere, Google).
* PromQL and other read APIs (including federation) (e.g. increased cumulative based operation accuracy).

Furthermore, it would unblock future Prometheus features for wider range of monitoring cases like:

* Delta temporality support.
* UpAndDown counter (i.e. not monotonic counters) e.g. StatsD.

> On top of that this work also allow to improve some existing features, notably various edge cases for native histogram resets hints (e.g. [TSDB chunk merges](https://github.com/prometheus/prometheus/issues/15346)). Arguably, CT can replace resent hint field in native histogram eventually.

Reviewer 2: ```suggestion
On top of that this work also allow to improve some existing features, notably various edge cases for native histogram resets hints (e.g. [TSDB chunk merges](https://github.com/prometheus/prometheus/issues/15346)). Arguably, CT can replace reset hint field in native histogram eventually.
```

* int64 Timestamp (T): When this value was observed.
* Labels: Unique identity of a series.
  * This includes special metadata labels like: `__name__`, `__type__`, `__unit__`
* Exemplars
* Metadata

Since the CT concept introduction in Prometheus we:

* Extended Prometheus protobuf scrape format to include CT per each cumulative sample (TODO link).
* Proposed (for OM 2) text format changes for CT scraping (improvement over existing OM1 `_created` lines) (TODO link).
* Expanded Scrape parser interface to return `CreatedTimestamp` per sample (aka per line).
* Optimized Protobuf and OpenMetrics parsers for CT use (TODO links).
* Implemented an opt-in, experimental [`created-timestamps-zero-injection`](https://prometheus.io/docs/prometheus/latest/feature_flags/#created-timestamps-zero-injection) feature flag that injects fake sample (V: 0, T: CT).
* Included CT in [Remote Write 2 specification](https://prometheus.io/docs/specs/prw/remote_write_spec_2_0/#ioprometheuswritev2request).

### Background: Delta temporality

See the details, motivations and discussions about the delta temporality in [PROM-48](https://github.com/prometheus/proposals/pull/48).

> The core TL;DR relevant for this proposal is that the delta temporality counter sample can be conceptually seen as a "mini-cumulative counter". Essentially delta is a single-sample (value) cumulative counter for a period between (sometimes inclusive sometimes exclusive depending on a system) start(ST)/create(CT) timestamp and a (end)timestamp (inclusive).

Author: It's mentioned "a delta start timestamp (ST)." in TL;DR, but you are right we could mention it clearly somewhere. Especially once we decide if we go with ST or CT fully, we could just use one and only have a quick mention about other. 👍🏽

In other words, instant query for `increase(<cumulative counter>[5m])` produces a single delta sample for a `[t-5m, t]` period (V: `increase(<counter>[5m])`, CT/ST: `now()-5m`, T: `now()`).

This proves that it's worth considering delta when designing a CT feature support.

### Background: CT (cumulative) vs ST (delta)

[Previous section](#background-delta-temporality) argues that conceptually the Cumulative Created Timestamp (CT) and Delta Start Timestamp (ST) are essentially the same thing. This is why typically they are stored in the same "field" in other system APIs and storages (e.g. start time in OpenTelemetry TODO link).

The notable difference when this special timestamp is used for cumulative vs delta samples is the dynamicity **characteristics** of this timestamp.

* For the cumulative we expect CT to change on every new counter restart, so:
  * Average: in the order of ~weeks/months for stable workloads, ~days/weeks for more dynamic environments (Kubernetes).
  * Best case: it never changes (infinite count) e.g days_since_X_total.
  * Worse case: it changes for every sample.

> * For the delta we expect CT to change for every sample.

Reviewer: @fionaliao Haven't we heard of cases where a delta doesn't have a CT? My understanding is that situation exists in the wild whether we want it to or not.  (Is it too late to drop deltas with no CT because that would break people?)

> * For the delta we expect CT to change for every sample.

Reviewer 3: Even without CTs, deltas could be used and produce useful results with the `sum_over_time()` function. In my opinion, dropping such values might be a step towards "perfect is the enemy of good" territory. 

Maybe a compromise could be to return a warning/info from a rate-like function about missing CTs and potentially inaccurate results?

  * created_timestamp SHOULD be provided for metrics that follow counter semantics (e.g. counters and histograms).
  * Receivers MAY reject those series without created_timestamp being set.

* Otel) More detailed, but still descriptive SHOULD rules only: In OpenTelemetry data model ([temporality section](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#sums:~:text=name%2Dvalue%20pairs.-,A%20time%20window%20(of%20(start%2C%20end%5D)%20time%20for%20which%20the,The%20time%20interval%20is%20inclusive%20of%20the%20end%20time.,-Times%20are%20specified)) CT is generally optional, but strongly recommended. Rules are also soft and we assume they are "SHOULD" because there is section of [handling overlaps](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#overlap). However, it provides more examples, which allow us to capture some specifics:
  * Time intervals are half open `(CT, T]`.
  * Generally, CT SHOULD:
    * `CT[i] < T[i]` (equal means unknown)
    * For a cumulative, CT SHOULD:
      * Be the same across all samples for the same count:

      ```
      CT[0], T[0]
      CT[0], T[1]
      CT[0], T[2]
      ```

>   * For delta, CT SHOULD (`=` because it looks the start time is exclusive):

Reviewer 2: ```suggestion
  * For delta, CT SHOULD (`=` because it looks like the start time is exclusive):
```

    * Remote Write ingestion would need to be persistent and stateful, which blocks horizontal scalability of receiving.
    * It limits effectiveness of using CT for PromQL operations like `rate`, `resets` etc.
    * It makes "rolloup" (write time recording rules that pre-calculate rates) difficult to implement.
  * Given immutability invariant (e.g. Prometheus), you can't effectively inject CT at a later time (out of order writes are sometimes possible, but expensive, especially for a single sample to be written in the past per series).
  * It's prone to OOO false positives (we ignore this error for CTs now in Prometheus).
  * It's producing an artificial sample, which looks like it was scraped.
* We can't implement delta temporarily effectively.

## Goals

* [MUST] Prometheus can reliably store, query, ingest and export cumulative created timestamp (CT) information (long term plan for [PROM-29](https://github.com/prometheus/proposals/blob/main/proposals/0029-created-timestamp.md#:~:text=For%20those%20reasons%2C%20created%20timestamps%20will%20also%20be%20stored%20as%20metadata%20per%20series%2C%20following%20the%20similar%20logic%20used%20for%20the%20zero%2Dinjection.))

> * [SHOULD] Prometheus can reliably store, query, ingest and export delta start time information. This unblocks [PROM-48 delta proposal](https://github.com/prometheus/proposals/pull/48). Notably adding delta feature later on should ideally not require another complex storage design or implementation.

Author: Where exactly in the current wording we have circular must goal?

The reason why I initially put SHOULD is that I can imagine a potential (low probability) reality where the solution for CT storage for cumulatives is considerably different than delta.

When that happens what we do? If we put MUST for deltas, we should NOT go this path, even if it makes sense to split solutions? That's why, from the perspective of finding the best solution for Prometheus, it feels it should be SHOULD, so balance and join solutions if we can.

However, et's change to MUST to prioritize delta -- sounds like the main reason to have MUST is to justify this effort for Grafana goals, and that's ok. 👍🏽 

Want to make that commit?

> * [SHOULD] Prometheus can reliably store, query, ingest and export delta start time information. This unblocks [PROM-48 delta proposal](https://github.com/prometheus/proposals/pull/48). Notably adding delta feature later on should ideally not require another complex storage design or implementation.

Author: Changed to MUST, since it's obviously what happened (:

* [SHOULD] Overhead of the solution should be minimal--initial overhead target set to maximum of 10% CPU, 10% of memory and 15% of disk space.
* [SHOULD] Improve complexity/tech-debt of TSDB on the way if possible.
* [SHOULD] Complexity of consuming CTs should be minimal (e.g. low amount of knowledge needed to use it).

## Non-goals

In this proposal we don't want to:

* Expand details on delta temporality. For this proposal it's enough to assume about delta, what's described in the [Background: Delta Temporality](#background-delta-temporality).
* Expand or motivate the non-monotonic counter feature.

## How

> Native CT support may feel like a big change with a lot of implications. However, we have all the technical pieces to do this, especially if we implement CT with

Author: Yes, discussed at the bottom and sooner we decide the better (: Resolving this comment.

(more or less) the same guarantees and details as normal sample timestamp (T), just optional.

General decisions and principles affecting smaller technical parts:

1. We propose to stick to the assumption that CT can change for every sample and don't try to prematurely optimize for the special cumulative best case. Rationales:

* This unlocks the most amount of benefits (e.g. also delta) for the same amount of work, it makes code simpler.

> * We don't know if we need special cumulative best case optimization (yet); also it would be also for some "best" cases. Once we know we can always add those optimizations.

Reviewer 2: ```suggestion
* We don't know if we need special cumulative best case optimization (yet); also, it would be for some "best" cases. Once we know we can always add those optimizations.
```

> * We don't know if we need special cumulative best case optimization (yet); also it would be also for some "best" cases. Once we know we can always add those optimizations.

Author: Yes, this is somewhat possible in case of errors or somebody duplicates streams of samples. Some backends override the old data that overlaps. Ideally we allow downstream backends to decide what to do, but for Prometheus we decided to not imply any validation by default at the moment and handle things on query time. 

All noted in `### Proposed ST semantics and validation`

2. Similarly, we propose to not have special CT storage cases per metric types. TSDB storage is not metric type aware, plus some systems allow optional CTs on gauges (e.g. OpenTelemetry). We propose keeping that storage flexibility.

3. We propose to treat CT as an *optional* data everywhere, simply because it's new and because certain metric type do not need it (gauges). For efficiency and consistency with scrape and Remote Write protocols we treat default value `int64(0)` as a "not provided" sentinel. This has a consequence of inability to provide the CT of an exact 0 value (unlikely needed and if needed clients needs to use 1m after or before.

4. Given the [CT and ST context](#background-ct-cumulative-vs-st-delta) the CT vs ST meaning blurs. We could rename CT feature now to ST to be consistent with other systems. We propose to stick to "created timestamp" (CT) naming in Prometheus instead of attempting to rename some or all uses to "start timestamp" (ST). The main reason being that CT and ST naming is equally "correct" on storage and given Prometheus is cumulative-first system, we stick to CT. See (and challenge!) the rationales in the [alternative](#rename-ct-to-st).

> Given (4) decision, below proposal only uses "CT" naming, but it also means "ST" in delta temporality context.

Author: Duplicate discussion, I acknowledge you are a bit upset for this first choice.

Let's discuss on https://github.com/prometheus/proposals/pull/60/files#r2344682987

Let's go through all the areas that would need to change:

### Feature flag `ct-storage`

To develop and give experimental state to users, we propose to add a new feature flag `created-timestamp-storage`. Similar to [exemplar-storage]https://prometheus.io/docs/prometheus/latest/feature_flags/#exemplars-storage) it will tell Prometheus to ingest CT and use new (potentially breaking compatibility) storage formats.

We propose to have a single flag for both WAL, Block storage, etc., to avoid tricky configuration.

> Notably, given persistence of this feature, similar to example storage, if users enabled and then disabled this feature, users will might be able to access their CTs through all already persistent pieces e.g. WAL).

Reviewer 2: ```suggestion
Notably, given persistence of this feature, similar to examplar storage, if users enable and then disable this feature, users might be able to access their CTs through all already persistent pieces e.g. WAL).
```

This feature could be considered to be switched to opt-out, only after it's finished (this proposal is fully implemented) stable, provably adopted and when the previous LTS Prometheus version is compatible with this feature.

### Proposed CT semantics and validation

See the [official CT semantics in the ecosystem](#background-official-ctst-semantics). In this section we propose what semantics we recommend (or enforce).

CT/ST notion was popularized by OpenTelemetry and early experience exposed a big challenge: CT/ST data is extremely unclean given early adoption, mixed instrumentation support and multiple (all imperfect) ways of ["auto-generation"](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstarttimeprocessor) (`subtract_initial_point` might the most universally "correct", but it added only recently). This means that handling (or reducing ) CT errors is an important detail for consumers of this data.

TODO: Just a draft, to be discussed.
TODO: There are questions around:

> * Should we do inclusive vs exclusive intervals?

Author: I'm leaving this open, but we agreed to keep the semantics out of this proposal. See ### Proposed ST semantics and validation

Given majority of our delta data is coming from OTel I'd lean towards Otel semantics, so `exclusive`

https://opentelemetry.io/docs/specs/otel/metrics/data-model/#:~:text=Delta%20temporality%20means%20that%20successive%20data%20points%20advance%20the%20starting%20timestamp.%20For%20example%2C%20from%20start%20time%20T0%2C%20delta%20data%20points%20cover%20time%20ranges%20(T0%2C%20T1%5D%2C%20(T1%2C%20T2%5D%2C%20(T2%2C%20T3%5D%2C%20and%20so%20on.

> * Given optionality of this feature, can we even reject sample on TSDB Append if CT is invalid? (MUST or SHOULD on interface?)

Author: Yes, all noted as so

  // types; rejecting such samples within the Request as a result.
  //
  // For Go, see github.com/prometheus/prometheus/model/timestamp/timestamp.go
  // for conversion from/to time.Time to Prometheus timestamp.
  //
  // Note that the "optional" keyword is omitted due to efficiency and consistency.
  // Zero value means value not set. If you need to use exactly zero value for
  // the timestamp, use 1 millisecond before or after.
  int64 created_timestamp = 3;
}
```

This has been partially approved and discussed in https://github.com/prometheus/prometheus/pull/17036

## Alternatives

### Rename CT to ST

See the [CT vs ST context](#background-ct-cumulative-vs-st-delta), but generally given we get close to OpenTelemetry
semantics, we could rename to "Start Time/Timestamp" (ST) naming in all elements.

Pros:
* Consistency with OpenTelemetry and Google
* There are more future users than previous users, explaining differences between CT and ST.

Cons:
* Technically, one could argue that both CT and ST naming are correct in this context. Prometheus is cumulative-first system, which means
  using CT would make sense.

> * We already have CT in scrape protocols and ~Remote Write 2.0; do we need to switch them too?

Author: ```suggestion
* We already have CT in scrape protocols (we go deeper into this in [OM 2.0](https://github.com/prometheus/docs/pull/2636)) and ~Remote Write 2.0; do we need to switch them too?
```

* Changing names often creates confusion
* Having a slight changed name already makes it clear that we talk about Prometheus semantics of the same thing

> **Rejected** due to not enough arguments for renaming (feel free to challenge this!).

Author: Yes! Thanks for mentioning. From notes and memory I think it was your vote to rename vs others to not rename or no opinion. 

This is a starting point for discussion, we can argument otherwise. Do you have more arguments towards renaming that would help motivate us?

When writing this I think I got sold on not renaming. Both are technically correct and given Prometheus being cumulative native it only makes sense to leave like this. This does not mean we shouldn't go against my vote if we have consensus here to rename.  How to find this consensus? (: What would convince you to keep `CT` naming? (: 

Do you mind helping with the vote if you are passionate to challenge this? What are we missing here?
